### PR TITLE
[Das Benchmarks]: Stress test 1 Full or Bridge Node against x Light Nodes

### PR DESCRIPTION
**Pull request by derrandz**
## Important

Merge when https://github.com/celestiaorg/celestia-node/pull/1376 is merged and `go.mod` is corrected to point to `celestia-node` main
## Overview

We are interested in benchmarking the bridge node against a multitude of light node groups, starting from (_say_) a 100 up to 100K

To do so, this PR adds test cases and local telemetry to enable the benchmark alongside with metrics collection for benchmarking results visualization.

More details on #79 

## Changes

- [x] Adds documentation for the new test plan 002
- [x] Adds first test case
- [x] Adds local compositions

- [x] Display how many light clients are currently connected to the bridge node (concurrency level)

              
- [x] And finally, since the charts are rough, we will work on smoothening them out (cleaning labels, improving readability and so on)

- [x] Adds k8s compositions?

- [ ] Add matrix for 64/128 square sizes ([_Check comment here_](https://github.com/celestiaorg/test-infra/pull/89#discussion_r997066530))
     - [x] 64
     - [ ] 128
- [ ] Add resource consumption metrics for the bridge node to analyze how # of light nodes impacts resource consumption
- [ ] Add comparative charts for results across multiple concurrency levels:
              example: DASing time for (100 light clients, 500 light clients, 1000 light clients) all in the same chart for comparison.
              
              
## How to run:

1. Start testground:
```bash
$ make tg-start
```

2. In another terminal start the telemetry infrastructure
```bash
$ make telemetry-infra-up
```

3. Run the test-case
```bash
$ make tg-run-composition RUNNER=local-docker TESTPLAN=das-benchmarks COMPOSITION=001-lights-dasing-latest-from-bridge-16-50-28
```

4. Go to `http://localhost:3000` to access Grafana
5. [Add prometheus as a data source](https://grafana.com/docs/grafana/latest/datasources/add-a-data-source/)
  5.1 For the URL, if you are running in a droplet, use your instance's IP instead of localhost

6. [Import the dashboard](https://rudimartinsen.com/2020/08/06/grafana-importing-dashboards/) under `./build/grafana/dashboards/benchmarks.json`

## Dependencies

This PR depends on https://github.com/celestiaorg/test-infra/issues/132 

**Owners:** @derrandz

### CODE REVIEW

File: docs/test-plans/002-DAS-Benchmarks/test-cases/tc-001-x-light-finish-das-before-block-time.md
Comment (Bidon15) [inline docs/test-plans/002-DAS-Benchmarks/test-cases/tc-001-x-light-finish-das-before-block-time.md:7]: It will be awesome to have 32/64/128 matrix as we are having an assumption that it works already here #96 😅

**Comment by derrandz** [comment]
To configure the celestia node with the benchmark parameters, these two PRs have to go in;

* https://github.com/celestiaorg/celestia-node/pull/1224: Makes `SampleAmount` configurable
* https://github.com/celestiaorg/celestia-node/pull/1225: Makes `SampleFrom` configurable

File: plans/002-das-benchmarks/tests/sync-latest/run_light.go
Comment (derrandz) [inline plans/002-das-benchmarks/tests/sync-latest/run_light.go:101]: @Bidon15 Can I get your guidance on this 🙏🏻 ?

**Comment by derrandz** [comment]
The  [DASer PR](https://github.com/celestiaorg/test-infra/pull/89#issuecomment-1290879818) is not required to tell configure the light node to start dasing from a specific height. We changed the logic of the test to unblock this PR, so that's no longer required.

The other one that's mentioned in the comment might be required down the line for different sample amounts.

**Comment by derrandz** [comment]
Define the Non Functional Requirements that have to be met for this test plan alongside the metrics to collect and their thresholds.

Referencing https://github.com/celestiaorg/test-infra/issues/108

**Comment by derrandz** [comment]
### Update

The course of this PR is changing to include test isolation from test setup, metrics' collection and the infrastructure to support the metrics collection.

Although hacky, it's convenient to take this route. After getting a fully working version, we will rewrite the history of this PR to clean this up.

**[More context in here](https://github.com/celestiaorg/test-infra/issues/108#issuecomment-1311565416)**

**Comment by derrandz** [comment]
Ongoing work to enable blackbox telemetry is in this [PR](https://github.com/celestiaorg/celestia-node/pull/1376/files)

**Comment by derrandz** [comment]
Progress update regarding the blackbox telemetry efforts:

We managed to get a few charts to look at, by which we measure the performance of the bridge node in terms of how well it’s serving the DASing process. At the moment, since we would benchmark a bridge node against a multitude of light nodes, we will go with the option of displaying charts per light node instance (_you can choose a random instance from the drop down in the screenshot_)

Aggregate charts that display the overall state of the DASing process across all light nodes is the next step. (_Check the PR's TODOs_)

Screen Shot 2022-11-23 at 00 27 18

_The Selection of Light Nodes from the dropdown_
Screen Shot 2022-11-23 at 00 27 24

**Comment by derrandz** [comment]
Some improvements to charting:

* Added the native influx-db metrics for testground to track # of alive light nodes (_a way to experiment with testground’s influxdb_)

* switched histogram chartings to use tabular like logic (see now DASing time and block time charts looking more clear per height)

Screen Shot 2022-11-24 at 00 39 38

**Review by Bidon15 — CHANGES_REQUESTED** [review]

File: compositions/cluster-k8s/das-benchmarks/64-square-size/001-lights-dasing-from-latest-bridges-1000.toml
Comment (derrandz) [reply compositions/cluster-k8s/das-benchmarks/64-square-size/001-lights-dasing-from-latest-bridges-1000.toml:17]: mistake, i thought it was in seconds. thanks for the catch!

File: tests/helpers/das-benchmarks/run_light.go
Comment (derrandz) [reply tests/helpers/das-benchmarks/run_light.go:87]: This is a relic from the make-it-work period! 
This IP is my instance's IP, and on that port lives the otel-collector. I will turn it into a configuration parameter and retrieve it from `runenv`.

Awesome catch!

File: compositions/cluster-k8s/das-benchmarks/64-square-size/001-lights-dasing-from-latest-bridges-800.toml
Comment (derrandz) [reply compositions/cluster-k8s/das-benchmarks/64-square-size/001-lights-dasing-from-latest-bridges-800.toml:20]: 7mil made the validator break at height 7 with error `request body too large`, let's discuss this in chat.

Comment (Bidon15) [reply compositions/cluster-k8s/das-benchmarks/64-square-size/001-lights-dasing-from-latest-bridges-800.toml:20]: Oh I see. We can trick by just exploding the amount of validators but still stick with 1 bridge node to comply with the benchmarking pre-reqs
wdyt?

File: go.mod
Comment (derrandz) [reply go.mod:8]: This is on purpose, that's where the blackbox instrumentation code exists.

**Comment by derrandz** [comment]
Some charts of percentage of the network and which height they were at
Screen Shot 2022-12-13 at 18 07 09

* Regular HW: https://snapshots.raintank.io/dashboard/snapshot/hAl8T5kiGhIQnMTajN4GYdiMhaX5tQkS

* Higher HW: https://snapshots.raintank.io/dashboard/snapshot/49UKmH3lVj0wrIP8YbFHeIh82XosLedC

**Comment by derrandz** [comment]
Having tested https://github.com/celestiaorg/celestia-node/pull/1376 on the temporary testing branch https://github.com/celestiaorg/test-infra/pull/140, the following error popped, which is caused by the new changes in #136 

```
Jan 15 21:16:30.133320  INFO    45.9675s    MESSAGE << validators[000] (d9bb59) >> detected data network IP: 16.78.1.25/16
Jan 15 21:16:30.239251  INFO    46.0736s    MESSAGE << validators[000] (d9bb59) >> Validator Received is equal to: 0
Jan 15 21:16:30.239366  INFO    46.0737s    MESSAGE << validators[000] (d9bb59) >> Randomized number is equal to: 0
Jan 15 21:16:30.239411  INFO    46.0737s    MESSAGE << validators[000] (d9bb59) >> Peers Range is equal to: 1
Jan 15 21:16:30.239469  INFO    46.0746s      OTHER << validators[000] (d9bb59) >> Iteration of i ->  1
Jan 15 21:16:30.239604  INFO    46.0739s      CRASH << validators[000] (d9bb59) >> goroutine 82 [running]:
runtime/debug.Stack()
        /usr/local/go/src/runtime/debug/stack.go:24 +0x65
github.com/testground/sdk-go/run.HandlePanics()
        /go/pkg/mod/github.com/testground/sdk-go@v0.3.0/run/panic.go:23 +0x39
panic({0x28699c0, 0xc0005d80d8})
        /usr/local/go/src/runtime/panic.go:884 +0x212
github.com/celestiaorg/test-infra/tests/helpers/common.GetRandomisedPeers(0x0, 0x1, {0x0, 0x0, 0x0})
        /plan/tests/helpers/common/validator.go:246 +0x1a5
github.com/celestiaorg/test-infra/tests/helpers/common.BuildValidator({0x37a7a30, 0xc002e79f80}, 0xc001b0a180, 0xc00105b3e0)
        /plan/tests/helpers/common/validator.go:227 +0x1539
github.com/celestiaorg/test-infra/tests/helpers/das-benchmarks.RunValidator(0xc001b0a180, 0xc00105b3e0)
        /plan/tests/helpers/das-benchmarks/run_validator.go:51 +0x357
github.com/celestiaorg/test-infra/tests/plans/das-benchmarks.LightsDasingLatest(0xc001b0a180, 0xc001b0a180?)
        /plan/tests/plans/das-benchmarks/001-lights-dasing-latest-from-bridge.go:14 +0xcc
github.com/testground/sdk-go/run.invoke.func4()
        /go/pkg/mod/github.com/testground/sdk-go@v0.3.0/run/invoker.go:173 +0x142
created by github.com/testground/sdk-go/run.invoke
        /go/pkg/mod/github.com/testground/sdk-go@v0.3.0/run/invoker.go:162 +0x78c
goroutine 1 [running]:
runtime/debug.Stack()
        /usr/local/go/src/runtime/debug/stack.go:24 +0x65
github.com/testground/sdk-go/runtime.(*RunEnv).RecordCrash(0xc001b0a180, {0x2536ec0?, 0xc0037fc090?})
        /go/pkg/mod/github.com/testground/sdk-go@v0.3.0/runtime/runenv_events.go:246 +0x7b
github.com/testground/sdk-go/run.invoke(0xc001b0a180, {0x25a29c0?, 0x33513b0})
        /go/pkg/mod/github.com/testground/sdk-go@v0.3.0/run/invoker.go:190 +0x8e5
github.com/testground/sdk-go/run.InvokeMap(0xc000006340?)
        /go/pkg/mod/github.com/testground/sdk-go@v0.3.0/run/invoker.go:77 +0x94
main.main()
        /plan/main.go:30 +0x25
```

Should be simple to resolve, will resolve later.

**Comment by derrandz** [comment]
[This](https://github.com/celestiaorg/test-infra/pull/89#issuecomment-1383256462) is resolved now by ignoring seeds nodes for this test case.

File: tests/helpers/das-benchmarks/run_validator.go
Comment (Bidon15) [inline tests/helpers/das-benchmarks/run_validator.go:91]: can be removed. Usually 1 minute is enough to get 3-4th block in tendermint

File: tests/plans/das-benchmarks/001-lights-dasing-latest-from-bridge.go
Comment (Bidon15) [inline tests/plans/das-benchmarks/001-lights-dasing-latest-from-bridge.go:23]: [blocking]: please wrap into a wrapperOnFailure func to actually signal earlier the sync-service to start terminating the test run if smt went sideways during the execution

**Review by Bidon15 — CHANGES_REQUESTED** [review]
👏

File: envvv
Comment (derrandz) [reply envvv:1]: lawl big miss

File: go.mod
Comment (derrandz) [reply go.mod:7]: unfortunately this is where the metrics work is at... we will need to wait til its merged before we revert it to point to main. Let's treat it as the line above it, celestia-app and keep using my fork ftm.

File: tests/helpers/common/validator.go
Comment (derrandz) [reply tests/helpers/common/validator.go:227]: Yes ser https://github.com/celestiaorg/test-infra/issues/145

File: tests/helpers/das-benchmarks/run_light.go
Comment (derrandz) [reply tests/helpers/das-benchmarks/run_light.go:88]: yesser

File: tests/helpers/das-benchmarks/run_light.go
Comment (derrandz) [reply tests/helpers/das-benchmarks/run_light.go:136]: removed

File: tests/plans/das-benchmarks/001-lights-dasing-latest-from-bridge.go
Comment (derrandz) [reply tests/plans/das-benchmarks/001-lights-dasing-latest-from-bridge.go:23]: not sure what you mean by `wrapperOnFailure`, do you mean `runenv.RecordFailure` ?

File: go.mod
Comment (Bidon15) [inline go.mod:7]: please use latest node version :D

File: tests/helpers/das-benchmarks/run_validator.go
Comment (Bidon15) [inline tests/helpers/das-benchmarks/run_validator.go:122]: where is signalAndWait or just barrier for finished state of the test?

File: tests/plans/das-benchmarks/001-lights-dasing-latest-from-bridge.go
Comment (Bidon15) [inline tests/plans/das-benchmarks/001-lights-dasing-latest-from-bridge.go:22]: if we caught the err and didn't signaled of finished state -> the test will be killed because of the daemon's runtime worker timeout. Please remember to wrap errors with a signal state

**Review by Bidon15 — CHANGES_REQUESTED** [review]

File: go.mod
Comment (derrandz) [reply go.mod:7]: I can't for the moment. This is based on the latest main. Check https://github.com/celestiaorg/test-infra/pull/89#discussion_r1080564470

File: Makefile
Comment (Bidon15) [inline Makefile:124]: Needs to be parked into a separate issue once we merge

File: build/docker-compose.yml
Comment (Bidon15) [inline build/docker-compose.yml:48]: Needs to be parked into a separate issue to port to cluster. Will do myself
cc: @smuu and @jrmanes

File: testkit/nodekit/node.go
Comment (Bidon15) [inline testkit/nodekit/node.go:23]: Can you elaborate why this is needed? 
I propose to revert this back to vanilla node's Types

In the future Super Light Nodes and Partial nodes will require us adding them here

**Review by Bidon15 — CHANGES_REQUESTED** [review]

File: Makefile
Comment (Bidon15) [reply Makefile:124]: #166

File: go.mod
Comment (Bidon15) [inline go.mod:6]: Can you cut a release on your fork(v0.7.0-black-box-metrics)? 🙏 
last thing we want is to be confused with the current way we see release numbers

File: Makefile
Comment (derrandz) [reply Makefile:124]: Thank you 🖖🏻

File: build/docker-compose.yml
Comment (derrandz) [reply build/docker-compose.yml:48]: thank you :shipit:

**Comment by derrandz** [comment]
Final nits tracked in an issue for future resolution https://github.com/celestiaorg/test-infra/issues/167

**Pull request merged**
